### PR TITLE
[node] ERC20 direct app installs

### PR DESCRIPTION
**Pull request by cf19drofxots**
This adds support for installing an app with deposits of different asset types (ETH and/or ERC20).

This only modifies non-virtual apps. A followup PR will extend this support for virtual apps.

### CODE REVIEW

**Comment by ArjunBhuptani** [comment]
Why does this modify only normal apps? How can we help you push this into virtual apps asap?

**Comment by cf19drofxots** [comment]
@ArjunBhuptani as @snario has mentioned to you we need to update the virtual apps to be able to use the `CoinTransfer` outcome (which is the work @IIIIllllIIIIllllIIIIllllIIIIllllIIIIll is doing, see [PR](https://github.com/counterfactual/monorepo/pull/1855)).

This new interpreter is required so that end parties can enter into an agreement with the intermediary (as specified in the protocol) with arbitrary assets (in this case either ETH or ERC20) and arbitrary amounts as opposed to the current implementation of using only the `TwoPartyFixedOutcomeFromVirtualAppETHInterpreter` which is fixed amounts and only ETH.

File: packages/node/src/models/proposed-app-instance-info.ts [hunk 1]
```diff
@@ -125,10 +132,15 @@ export class ProposedAppInstanceInfo implements AppInstanceInfo {
...
+      // the AppInstance ID so there won't be a possible mismatch between
+      // a proposed AppInstance ID and an installed AppInstance ID
       {
-        limit: bigNumberify(this.myDeposit).add(this.peerDeposit),
-        tokenAddress: CONVENTION_FOR_ETH_TOKEN_ADDRESS
-      }
+        limit: [],
+        tokens: [],
+        transferTokenIndexes: []
+      } as CoinTransferInterpreterParams
```
[the remark this answers is not in the document]
Comment (cf19drofxots) [reply packages/node/src/models/proposed-app-instance-info.ts]: I think it's natural for an `AppInstance` to know what parameters it was installed with.

File: packages/node/src/models/app-instance.ts [hunk 2]
```diff
@@ -44,11 +42,9 @@ export type AppInstanceJson = {
   coinTransferInterpreterParams?: {
     // Derived from:
     // packages/contracts/contracts/interpreters/CoinTransferETHInterpreter.sol#L18
-    limit: { _hex: string };
-    tokenAddress: string;
+    limit: { _hex: string }[];
+    tokens: string[];
```
[the remark this answers is not in the document]
Comment (cf19drofxots) [reply packages/node/src/models/app-instance.ts]: Yeah though to keep encoding/decoding straightforward `tokenAddress` would also need to be used on the contracts

File: packages/types/src/data-types.ts [hunk 3]
```diff
@@ -104,6 +113,13 @@ export enum TwoPartyFixedOutcome {
   SPLIT_AND_SEND_TO_BOTH_ADDRS = 2
 }
 
+export type CoinBalanceRefundState = {
+  recipient: string;
+  multisig: string;
+  threshold: BigNumber;
+  token: string;
```
Comment (snario) [inline packages/types/src/data-types.ts]: `tokenAddress`

**Comment by cf19drofxots** [comment]
Now that this is in, I'll add a PR to address this issue: https://github.com/counterfactual/monorepo/issues/1834

**Comment by ArjunBhuptani** [comment]
Can we prioritize uninstall bug instead (critical)? I believe the issue in #1834 is an edge case which we can handle as a part of test week

File: packages/high-roller-bot/package.json [hunk 4]
```diff
@@ -33,8 +33,8 @@
   },
   "dependencies": {
     "@counterfactual/firebase-client": "0.0.3",
-    "@counterfactual/node": "0.2.17",
-    "@counterfactual/types": "0.0.22",
+    "@counterfactual/types": "0.2.18",
```
Comment (cf19drofxots) [inline packages/high-roller-bot/package.json]: This was updated wrongly unfortunately.
